File: Source Code/Samples/Bölüm 16/Delegeler/sample.vb
Imports System.Threading
Imports System.IO
Imports System.Net


Class Program
    Delegate Function GetWebResponse() As WebResponse
    Shared sync As Object = New Object

    Shared tamamlanan As Integer = 0

    Shared Sub Main(ByVal args As String())
        Dim req As HttpWebRequest
        Dim resp As GetWebResponse
        Dim callBack As AsyncCallback
        For i As Integer = 0 To 9

            req = CType(HttpWebRequest.Create(("http://www.xxx.com/indir.aspx?Id=" + i.ToString)), HttpWebRequest)
            resp = AddressOf req.GetResponse
            callBack = New AsyncCallback(AddressOf IslemTamamlandi)
            resp.BeginInvoke(callBack, resp)
        Next
        Console.WriteLine("Ýþlemin tamamlanmasý bekleniyor ...")
        SyncLock sync
            While (tamamlanan < 10)
                Thread.Sleep(10)
                Monitor.Wait(sync)

            End While

        End SyncLock
        Console.WriteLine("Ýþlem tamamlandý")
    End Sub

    Private Shared Sub IslemTamamlandi(ByVal sonuc As IAsyncResult)
        Console.WriteLine("{0} numaralý ii parçacýðý tamamlandý ...", Thread.CurrentThread.ManagedThreadId)
        Dim r As GetWebResponse = CType(sonuc.AsyncState, GetWebResponse)
        Try
            Using resp As WebResponse = r.EndInvoke(sonuc)
                Using reader As StreamReader = New StreamReader(resp.GetResponseStream)
                    Console.WriteLine(reader.ReadToEnd)
                    reader.Close()
                End Using
                resp.Close()
            End Using
        Catch exc As Exception
            Console.WriteLine(exc.Message)
        End Try
        SyncLock sync
            tamamlanan = tamamlanan + 1
            Monitor.Pulse(sync)

        End SyncLock
    End Sub


End Class

File: Source Code/Samples/Bölüm 16/Delegeler/sample.cs
using System;
using System.Threading;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Net;


namespace DotNetKitabý
{
    class Program
    {

        delegate WebResponse GetWebResponse();
        static object sync = new object();
        static int tamamlanan = 0;
        public static void Main()
        {
            HttpWebRequest req;
            GetWebResponse resp;
            AsyncCallback callBack;
            for (int i = 0; i < 10; i++)
            {
                req = (HttpWebRequest)HttpWebRequest.Create("http://www.xxx.com/indir.aspx?Id=" + i.ToString());
                resp = req.GetResponse;

                callBack = new AsyncCallback(IslemTamamlandi);
                Thread.Sleep(10);
                resp.BeginInvoke(callBack, resp);
            }

            Console.WriteLine("Ýþlemin tamamlanmasý bekleniyor ...");

            lock (sync)
            {
                while (tamamlanan < 10)
                {
                    Thread.Sleep(10);
                    Monitor.Wait(sync);
                }
            }

            Console.WriteLine("Ýþlem tamamlandý");

            Console.ReadLine();
        }

        static void IslemTamamlandi(IAsyncResult sonuc)
        {
            Console.WriteLine("{0} numaralý iþ parçacýðý tamamlandý ...", Thread.CurrentThread.ManagedThreadId);
            GetWebResponse r = (GetWebResponse)sonuc.AsyncState;

            try
            {
                using (WebResponse resp = r.EndInvoke(sonuc))
                {
                    using (StreamReader reader = new StreamReader(resp.GetResponseStream()))
                    {
                        Console.WriteLine(reader.ReadToEnd());
                        reader.Close();
                    }
                    resp.Close();
                }
            }
            catch (Exception exc)
            {
                Console.WriteLine(exc.Message);
            }

            lock (sync)
            {
                tamamlanan++;
                Monitor.Pulse(sync);
            }

        }



    }


}

